src/common/py-serde/src/python.rs

```diff
@@ -71,25 +69,28 @@ macro_rules! impl_bincode_py_state_serialization {
         #[cfg(feature = "python")]
         #[pymethods]
         impl $ty {
-            pub fn __reduce__(&self, py: Python) -> PyResult<(PyObject, PyObject)> {
-                use pyo3::types::PyBytes;
-                use pyo3::PyTypeInfo;
-                use pyo3::ToPyObject;
+            pub fn __reduce__<'py>(
+                &self,
+                py: Python<'py>,
+            ) -> PyResult<(PyObject, (pyo3::Bound<'py, pyo3::types::PyBytes>,))> {
+                use pyo3::{
+                    types::{PyAnyMethods, PyBytes},
+                    PyTypeInfo, ToPyObject,
+                };
                 Ok((
-                    Self::type_object(py)
+                    Self::type_object_bound(py)
                         .getattr("_from_serialized")?
-                        .to_object(py),
-                    (PyBytes::new(py, &$crate::bincode::serialize(&self).unwrap()).to_object(py),)
-                        .to_object(py),
+                        .into(),
+                    (PyBytes::new_bound(
+                        py,
+                        &$crate::bincode::serialize(&self).unwrap(),
+                    ),),
                 ))
             }
 
             #[staticmethod]
-            pub fn _from_serialized(py: Python, serialized: PyObject) -> PyResult<Self> {
-                use pyo3::types::PyBytes;
-                serialized
-                    .extract::<&PyBytes>(py)
-                    .map(|s| $crate::bincode::deserialize(s.as_bytes()).unwrap())
+            pub fn _from_serialized(serialized: &[u8]) -> Self {
```

Comment:
maybe this is a good contender for the bound api so we don't have to `Clone / Copy` the underlying bytes before passing them into rust. We could instead of a GILBound to the bytes and create the underlying rust obj

Reply:
I'm not entirely sure what you mean. When a pyfunction parameter is of type `&[u8]`, pyo3 just calls `.as_bytes()` internally and returns a reference to the actual bytes in the python object without copying.